attribute.go

```diff
@@ -301,6 +307,32 @@ func (ad *AttributeDecoder) Do(fn func(b []byte) error) {
 	}
 }
 
+// Nested decodes data into a nested AttributeDecoder and invokes fn repeatedly
+// for each attribute found in the nested data. When calling Nested, the Next
+// and Err methods should not be called on the nested AttributeDecoder.
+//
+// The nested AttributeDecoder nad inherits the same ByteOrder setting as the
+// top-level AttributeDecoder ad.
+func (ad *AttributeDecoder) Nested(fn func(nad *AttributeDecoder) error) {
+	// Because we are wrapping Do, there is no need to check ad.err immediately.
+	ad.Do(func(b []byte) error {
+		nad, err := NewAttributeDecoder(b)
+		if err != nil {
+			return err
+		}
+		nad.ByteOrder = ad.ByteOrder
+
+		// Iterate until no more attributes or nad.Err() != nil.
+		for nad.Next() {
```

Comment:
I'm not sure embedding the Next loop is a good idea because it makes certain operations harder. For example:

```go
// Array of peers.
			ad.Nested(func(nad *netlink.AttributeDecoder) error {
				// Individual peer fields are parsed from another nested set
				// of attributes within the peers array.
				var p wgtypes.Peer
				nad.Nested(func(nnad *netlink.AttributeDecoder) error {
					return parsePeer(nnad, &p)
				})

				// Only append this peer if no problems occurred.
				if nad.Err() != nil {
					return err
				}

				d.Peers = append(d.Peers, p)
				return nil
			})
```

The parsePeer function is now called repeatedly and only parses a single field at a time, which is a bit awkward.

Reply:
it might also lead to hard to read code, if there are multiple level of nestings.

e.g. in case of conntrack, the attribute CTA_IP_V6_SRC is nested in CTA_TUPLE_IP, which is nested in CTA_TUPLE_ORIG.

Reply:
I changed this so that the caller is still responsible for calling ad.Next. As far as multiple levels of nesting, I think it's up to the caller to use good judgment and refactor out heavily indented blocks.